V2 tracker module [WIP]

anleo opened the pull request.

anleo committed:
- tracker module WIP

anleo committed:
Merge branch 'v2' into v2-tracker-module-merge

anleo committed:
- tracker module WIP

irinadrozdova commented on public-v2/src/app/tracker/services/task.service.ts:
```diff
@@ -9,6 +9,17 @@ export class TaskService {
+    return this.taskResource.getChildrenTasks({}, {taskId: taskId})
+      .$observable
+      .map((tasks: Task[]) => {
+        return tasks;
```
не понятно зачем map нужен ?

irinadrozdova commented on public-v2/src/app/tracker/services/task.service.ts:
```diff
@@ -20,6 +31,28 @@ export class TaskService {
+    return this.taskResource.getTask({}, {taskId: taskId})
+      .$observable
+      .map((task: Task) => {
+        return task;
```
тоже не пойму зачем map ?)

irinadrozdova commented on public-v2/src/app/tracker/resolvers/task.resolver.ts:
```diff
@@ -0,0 +1,20 @@
+  }
+
+  resolve(route: ActivatedRouteSnapshot): Observable<Task> {
+    return this.taskService.getTask('' + route.params['taskId'])
```
вынести приведение к стрингу в метод getTask

irinaudovyk replied:
здесь тоже попробуй без ' ' +, должно работать;

irinaudovyk commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+    let tasks = this.tasks;
+    this.tasks = [];
+
+    if ('' + task._id === '' + this.task._id) {
```
думаю, лучше использовать .toString() для приведения к строке

irinadrozdova replied:
попробуй без приведения к строке

irinaudovyk commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+      this.task = task;
+    } else if (tasks.find((_task) => _task._id === task._id)) {
+      tasks = tasks.map((_task) => {
+        return ('' + _task._id === '' + task._id) ? task : _task;
```
.toString()

irinadrozdova replied:
попробуй без приведения к строке )

irinadrozdova commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+import {Task} from '../../models/task';
+import {TaskService} from "../../services/task.service";
+import {ActivatedRoute, Params} from "@angular/router";
+import {Observable} from "rxjs";
```
библиотеки и не библиотеки, Андрюха меня понимает ;)

irinadrozdova commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+      .subscribe(task => {
+        this.parentTask = null;
+        this.root = null;
+        this.tasks = [];
```
у тебя уже изначально parentTask и root  = null

anleo replied:
обнуляю при смене роута (таска) без этого не получается правильно обновлять страницу

irinadrozdova commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+  parentTask: Task|null = null;
+  root: Task|null = null;
+  editTask: Task|null = null;
+  tasks: Task[] = [];
```
возможно все эти переменные вынести в какой-нибудь класс модельки ? TaskItem, возможно тебе будет так чуток легче, так как меньше букв будет

irinadrozdova commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+
+  loadTasks(taskId: string): Observable<Task[]> {
+    if (taskId) {
+      return this.taskService.getChildrenTasks(taskId);
```
может тут написать просто getTasks? 
внутри метода уже решать, если пришел id, то по id, иначе просто забери все.
и тогда тебе не нужен будет getChildrenTasks

irinaudovyk commented on public-v2/src/app/tracker/services/task.service.ts:
```diff
@@ -9,6 +9,17 @@ export class TaskService {
+      .map((tasks: Task[]) => {
+        return tasks;
+      })
+      .catch((err) => {
```
ошибки проще ловить в subscribe, a здесь можно обойтись без catch

irinaudovyk commented on public-v2/src/app/tracker/resources/tasks.resource.ts:
```diff
@@ -33,4 +39,20 @@ export class TaskResource extends Resource {
   getTasks: ResourceMethod<any, Task[]>;
+
+  @ResourceAction({
+    path: '/{taskId}'
```
лучше перестраховаться и писать {!taskId}

irinadrozdova commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+    if ('' + task._id === '' + this.task._id) {
+      this.task = task;
+    } else if (tasks.find((_task) => _task._id === task._id)) {
+      tasks = tasks.map((_task) => {
```
перебор с поиском таска )  выглядит как мап лишний )
и чуток не понятно, что должно произойти в else if ,  если просто в else пушим данный таск

irinadrozdova commented on public-v2/src/app/tracker/tasks/task-panel/task-panel.component.html:
```diff
@@ -13,7 +13,7 @@
     <h4 class="panel-title">
       <!--<a ui-sref="app.task({taskId: task._id})" data-nodrag>{{task.title}}</a>-->
-      <a [routerLink]="[task._id]">{{task.title}}</a>
+      <a [routerLink]="['/app/tasks/' + task._id]">{{task.title}}</a>
```
[routerLink]="['/app/tasks/' +  task._id], вместо + должна быть ,

irinaudovyk commented on public-v2/src/app/tracker/tasks/task-item/task-item.component.ts:
```diff
@@ -0,0 +1,86 @@
+  ngOnInit() {
+    this.route.params
+      .switchMap((params: Params) => {
+        return this.taskService.getTask(params['taskId']);
```
task уже забрал resolver, здесь его можно забрать через snapshot

anleo replied:
не получается перегружать таск, на одном и том же уровне, зарезолвился один таск и он не меняется.

anleo committed:
- fixed answer on delete from server
- fixed order of libs
- simplified requests to taskService
- fixed resource paths
- updated edit mode
